### PR TITLE
[16.0][MIG] website_sale_coupon_selection_wizard>website_sale_loyalty_suggestion_wizard: Migration to version 16.0

### PR DESCRIPTION
**Pull request by pilarvargas-tecnativa**
cc @Tecnativa TT44379

@chienandalu please review

### CODE REVIEW

**Comment by pilarvargas-tecnativa** [comment]
@chienandalu I think it should be renamed to website_sale_loyalty_order_suggestion, since the most direct dependency is to sale_loyalty_order_suggestion, what do you think about it?

File: sale_loyalty_order_suggestion/wizard/sale_loyalty_reward_wizard.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ def _compute_loyalty_rule_line_description(self):
             and not self.applicable_program
             and units_required > 0
         ):
-            products = self.loyalty_rule_line_ids.mapped("product_id.name")
+            products = self.loyalty_rule_line_ids.mapped("product_id.display_name")
```
Comment (pilarvargas-tecnativa) [inline sale_loyalty_order_suggestion/wizard/sale_loyalty_reward_wizard.py]: Note for me: Remove this change in this PR

**Comment by pilarvargas-tecnativa** [comment]
@carlos-lopez-tecnativa can you take a look please?

**Comment by carlos-lopez-tecnativa** [comment]
> @carlos-lopez-tecnativa can you take a look please?

@pilarvargas-tecnativa The Runboat is pending. Could you please rebase the branch?

**Comment by pilarvargas-tecnativa** [comment]
> @pilarvargas-tecnativa The Runboat is pending. Could you please rebase the branch?

Done!

File: website_sale_loyalty_suggestion_wizard/templates/promotion_templates.xml [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        active="True"
+        customize_show="True"
+        name="Promotions suggestions"
+    >
+        <xpath expr="//td[hasclass('td-img')]" position="before">
+            <t
+                t-set="promo"
+                t-value="line.suggested_promotion_ids.filtered(lambda x: x.is_published)[:1].id"
+            />
+            <td class="td-promo-suggestion" t-if="promo">
```
Comment (carlos-lopez-tecnativa) [inline website_sale_loyalty_suggestion_wizard/templates/promotion_templates.xml]: Functionally tested and working, but this new conditional `<td>` breaks the UI. Please ensure a corresponding `<td>` is added in the header as well.

image

Comment (pilarvargas-tecnativa) [reply website_sale_loyalty_suggestion_wizard/templates/promotion_templates.xml]: Done, thanks!

image

File: website_sale_loyalty_suggestion_wizard/templates/promotion_templates.xml [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                            <div>
+                                <strong t-out="promo.public_name" />
+                            </div>
+                        </td>
+                        <td />
+                        <td class="w-25 text-center">
+                            <a
+                                role="button"
+                                class="btn btn-info"
+                                t-att-href="'/promotions/%s/apply' % promo.id"
```
Comment (carlos-lopez-tecnativa) [inline website_sale_loyalty_suggestion_wizard/templates/promotion_templates.xml:180]: This URL is not working
image

image

**Review by carlos-lopez-tecnativa — CHANGES_REQUESTED** [review]

File: website_sale_loyalty_suggestion_wizard/templates/promotion_templates.xml (hunk 3 above)
Comment (pilarvargas-tecnativa) [reply website_sale_loyalty_suggestion_wizard/templates/promotion_templates.xml:180]: In the last change I had deleted a necessary variable, it is now restored.

File: website_sale_loyalty_suggestion_wizard/templates/promotion_templates.xml [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                t-set="promotions"
+                t-value="website_sale_order.website_order_line.suggested_promotion_ids.filtered(lambda x: x.is_published)[:1].id"
+            />
+            <t t-if="promotions">
+                <th />
+            </t>
+        </xpath>
+        <xpath expr="//td[hasclass('td-img')]" position="before">
+            <td class="td-promo-suggestion" t-if="promotions">
+                <t t-set="promo" t-value="line.suggested_promotion_ids[:1].id" />
```
Comment (carlos-lopez-tecnativa) [inline website_sale_loyalty_suggestion_wizard/templates/promotion_templates.xml]: **I tested this case:**

I created two rules: one for the Storage Box and another for the Large Cabinet.
Both products are displayed on the website.

- Add both products to the cart.
- Go to the cart, and the first promotion sale-promotion (related to the Storage Box) is suggested. Close the wizard.
- In the cart, on the line with the Large Cabinet product, click the button to apply the promotion.
- The wizard displays the promotion related to the Storage Box product, not the Large Cabinet.

I expected the wizard to display the promotion related to the selected product.
I’ve attached a video for more details. Is this behavior expected?
sale-promotion

**Review by carlos-lopez-tecnativa — COMMENTED** [review]
`

File: website_sale_loyalty_suggestion_wizard/templates/promotion_templates.xml (hunk 4 above)
Comment (pilarvargas-tecnativa) [reply website_sale_loyalty_suggestion_wizard/templates/promotion_templates.xml]: It should be resolved by now

**Review by carlos-lopez-tecnativa — COMMENTED** [review]
I found a new mismatch. Please let me know if this is an issue or expected behavior:

I created two rules: one for the Storage Box and another for the Large Cabinet.
Both products are displayed on the website.
Add both products to the cart.
Clear any applied promotions.
Apply the first promotion. The button to apply the promotion on the corresponding line becomes hidden, which is correct.
Remove the line with the applied promotion.
Apply the second promotion. However, the button to apply the promotion remains visible when it should be hidden, as it was for the first line.
If the user clicks the promotion button in this state, the URL is set to "False" (as mentioned in my previous comment) and results in a 404 error.

sale-promotion

**Comment by pilarvargas-tecnativa** [comment]
> I found a new mismatch. Please let me know if this is an issue or expected behavior:

Fixed, the error was causing that there is another promotion not published, before making the change to adapt the columns of the table that cell was conditioned to promo, when making the change to adapt the size of the table was missing add that condition to the cell of the icon.

**Review by carlos-lopez-tecnativa — APPROVED** [review]
Tested functionality, LGTM.

**Comment by pedrobaeza** [comment]
@chienandalu please confirm it's OK

**Review by chienandalu — APPROVED** [review]
Thanks :)

**Pull request merged**